store/postgres/migrations/2019-02-26-182914_remove_unassigned_deployment_indexes_procedure/up.sql

```diff
@@ -0,0 +1,41 @@
+-- Find all subgraphs without deployment assignments and remove their indexes
+CREATE OR REPLACE FUNCTION remove_unassigned_deployment_indexes()
+    RETURNS VOID AS
+$$
+DECLARE
+    index_to_drop RECORD;
+    counter INT := 0;
+BEGIN
+    -- Loop through each index
+    FOR index_to_drop IN
+        -- Get all indexes on subgraphs without deployment assignments
+        SELECT
+          indexname as name
+        from pg_indexes
+        where
+          left(indexname,46) IN
+          (
+            select left(lower(deployments.id),46)
+            from entities as deployments
+            left join
+              (
+                select
+                  id, entity
+                from entities
+                where
+                  subgraph='subgraphs' and
+                  entity='SubgraphDeploymentAssignment'
+              ) assigments
+            on deployments.id=assigments.id
+            where
+              deployments.subgraph='subgraphs' and
+              deployments.entity='SubgraphDeployment' and
+              assigments.id is null)
```

Comment:
I am not clear on what can go wrong if assignments are being modified concurrently, like is it possible that we are in the process of setting up an assignment, have already created the indexes, but not updated the assignments. Do we need to lock the assignments rows `for update` while we do this ?

Reply:
Scratch the thing about 'assignments rows', there's nothing to lock there. If assignments was its own table, I'd take an exclusive lock on the table to bar any concurrent modifications from it; but taking an exclusive lock on entities might be a bit too much.

Reply:
Good point, I'll do some research into the `SubgraphDeploymentAssignment`operations to ensure that we don't delete indexes for a subgraph that is in the process of being deployed.

Reply:
@lutter It looks like there is no risk of deleting a subgraph's indexes before it is given a deployment assignment because the assignment happens before the subgraph is started and its indexes are created.  The indexes are built in `provider.start()` which is called using an existing `SubgraphDeploymentAssignment` (if the query fails to get an existing assignment an error is triggered.)  https://github.com/graphprotocol/graph-node/blob/master/core/src/subgraph/registrar.rs#L194-L215